lib/mongo/server/connection.rb

```diff
@@ -190,9 +190,9 @@ def handshake!
 
       def authenticate!
         if options[:user] || options[:auth_mech]
-          user = Auth::User.new(Options::Redacted.new(:auth_mech => default_mechanism).merge(options))
+          @user ||= Auth::User.new(Options::Redacted.new(:auth_mech => default_mechanism).merge(options))
```

Comment:
We unfortunately can't cache the user and must set it each time we connect because the auth mechanism could have changed. This commit describes the scenario: https://github.com/mongodb/mongo-ruby-driver/commit/49f534668d1a4d32c44d300557da59fbd4765d0f